File: Compression.py
import string
import re


class Compress:
    def __init__(self, books):
        self.books = books
        self.dictionary = {}
        self.string_dict = bytearray()
        self.pointers = []
        self.ptr_to_post = []
        self.words = []
        self.__create()

    def __create(self) -> None:
        file_id = 1
        for i in self.books:
            try:
                with open(i) as file:
                        for line in file:
                            for word in re.findall(
                                    r"[^—…”“‘" + string.whitespace + string.punctuation + string.digits + "]+", line):
                                word = word.lower()
                                if word in self.dictionary:
                                    if file_id not in self.dictionary[word]:
                                        self.dictionary[word].append(file_id)
                                else:
                                    self.dictionary[word] = [file_id]
                file_id += 1
            except FileNotFoundError:
                print('Файл {} не знайдено!'.format(i))
                del self.books[i]
        self.words = list(self.dictionary.keys())
        self.words.sort()

    def compress_dict(self):
        k = 0
        for w in self.words:
            if k % 4 == 0:
                curr_length = len(self.string_dict)
                self.pointers.append(curr_length)
            self.string_dict.append(len(w))
            self.string_dict += w.encode()
            k += 1

    def compress_postings(self):
        for k in range(len(self.words)):
            self.ptr_to_post.append(None)
        # print(len(self.words))
        for k in range(len(self.words)):
            v = self.dictionary[self.words[k]]
            # print(self.words[k])
            v = Compress.differences(v)
            numb = "1"
            for i in v:
                g = Compress.gamma(i)
                numb += g
            dec = int(numb, 2)
            print(numb)
            # print(dec)
            self.ptr_to_post[k] = dec
            print(Compress.decode_gamma(dec))

    @staticmethod
    def gamma(n) -> str:
        b = str(bin(n))[3:]
        unar = '1'*len(b)
        return unar+'0'+b

    @staticmethod
    def decode_gamma(g) -> list:
        b = str(bin(g))[3:]
        res = []
        ch = 0
        count = 0
        n = '1'
        while ch < len(b):
            if b[ch] == '1':
                count += 1
                ch += 1
            elif b[ch] == '0':
                for i in range(count):
                    ch += 1
                    n += b[ch]
                res.append(int(n, 2))
                n = '1'
                ch += 1
                count = 0
        return res

    @staticmethod
    def differences(posting_list: list) -> list:
        diffs = [posting_list[0]]
        for i in range(len(posting_list) - 1):
            diffs.append(posting_list[i + 1] - posting_list[i])
        return diffs

    def write(self) -> None:
        with open('dict.txt', 'wb') as file:
            file.write(self.string_dict)
            # file.write('{}\n{}'.format(self.string_dict, self.pointers))

File: Compression.js
const fs = require('fs');

class Compression {
    constructor(books) {
        this.books = books;
        this.dictionary = new Map();
        this.stringDict = '';
        this.pointers = [];
        this.ptrToPost = [];
        this.words = [];
        this.__init();
    }

    compressDict() {
        let blockLen = 0;

        for (let i = 0; i < this.words.length; i++) {
            if (blockLen % 4 === 0) {
                this.pointers.push(this.stringDict.length);
            }
            this.stringDict += `${this.words[i].length}${this.words[i]}`;
            this.blockLen++;
        }
        fs.writeFileSync(outputPath, this.stringDict);
    }

    compressPostings() {
        const wstream = fs.createWriteStream('myBinaryFile.txt');
        // creates random Buffer of 100 bytes
        // var buffer = crypto.randomBytes(100);

        // wstream.end();
        // this.ptrToPost = new Array(this.words.length).fill();
        // console.log(this.dictionary.get(this.words[1]));
        for (let i = 0; i < /* this.words.length */ 20; i++) {
            const docIds = this.dictionary.get(this.words[i]);
            // console.log(this.words[i]);
            console.log({ docIds });
            const distances = this.__distance(docIds);
            console.log({ distances });
            let numb = '1';
            distances.map(dist => numb += this.__gammaEncode(dist));
            // for (let j = 0; j < distances.length; j++) {
            //     numb += this.__gammaEncode(distances[j]);
            // }
            console.log({ numb });
            const decimalNumb = parseInt(numb, '2');
            this.ptrToPost.push(decimalNumb);
            // console.log({ decimalNumb });
            // console.log(this.__gammaDecode(numb));
        }
        wstream.write(Buffer.from(this.ptrToPost));
        wstream.end();
        // console.log();
    }

    __init() {
        for (let i = 0; i < this.books.length; i++) {
            const data = fs.readFileSync(this.books[i], 'utf-8');
            const words = data.replaceAll('\n', ' ')
                // .replaceAll(/[^a-zA-Z ]+/g, '')
                .replaceAll(/[^а-яА-Я ]+/g, '')
                .split(' ');

            const docId = i + 1;
            for (let j = 0; j < words.length; j++) {
                if (!!words[j]) {
                    const word = words[j].toLowerCase();
                    const docIds = this.dictionary.get(word);
                    if (docIds) {
                        if (!docIds.includes(docId)) docIds.push(docId);
                        this.dictionary.set(word, docIds);
                    } else {
                        this.dictionary.set(word, [docId]);
                    }
                }
            }
        }
        // console.log(this.dictionary);
        this.words = [...this.dictionary.keys()];
        this.words.sort();
        // console.log([...this.dictionary.values()].filter(x => x.length > 1));
    }

    __gammaEncode(n) {
        const binary = n.toString('2').slice(1);
        const unary = `${new Array(binary.length).fill(1).join('')}0`;
        return `${unary}${binary}`;
    }

    /**
     * 
     * @param {string} gamma binary number
     * @returns array of decimals
     */
    __gammaDecode(gamma) {
        const binary = gamma.slice(1);
        const numbers = []
        let ch = 0
        let ctr = 0
        let n = '1'

        while (ch < binary.length) {
            if (binary[ch] === '1') {
                ctr += 1;
                ch += 1;
            } else if (binary[ch] === '0') {
                for (let i = 0; i < ctr; i++) {
                    ch++;
                    n += binary[ch];
                }
                numbers.push(parseInt(n, 2));
                n = '1';
                ch++;
                ctr = 0;
            }
        }
        return numbers;
    }

    __distance(docIds) {
        const dist = [docIds[0]];
        for (let i = 0; i < docIds.length - 1; i++) {
            dist.push(docIds[i + 1] - docIds[i]);
        }
        return dist;
    }
}

module.exports = {
    Compression,
}